### PR TITLE
Add is_unique_sorted and unique_sorted

### PR DESCRIPTION
**Pull request by thibault-martinez**
## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

**Review by Adam-Gleave — APPROVED** [review]
Looks good to me!

File: bee-common/bee-common/src/ord.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! A module that provides features related to ordering.
+
+use core::{cmp::Ordering, iter::Iterator};
+
+/// Checks if an iterator yields ordered and unique values.
+pub fn is_unique_sorted<T: Ord, I: Iterator<Item = T>>(mut iterator: I) -> bool {
```
Comment (zesterer) [inline bee-common/bee-common/src/ord.rs:9]: How does this code check for uniqueness?

File: bee-common/bee-common/src/ord.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    let mut previous = match iterator.next() {
+        Some(e) => e,
+        None => return true,
+    };
+
+    for curr in iterator {
+        if previous.cmp(&curr) != Ordering::Less {
```
Comment (zesterer) [inline bee-common/bee-common/src/ord.rs:16]: Why not `iter.is_sorted_by(|a, b| a.cmp(b).filter(|o| o == &Ordering::Greater))` instead?

Comment (thibault-martinez) [reply bee-common/bee-common/src/ord.rs:16]: It's nightly only

File: bee-common/bee-common/src/ord.rs (hunk 1 above)
Comment (thibault-martinez) [reply bee-common/bee-common/src/ord.rs:9]: It does it by checking order AND uniqueness at the same time, two consecutive items can't be equal

File: bee-common/bee-common/src/ord.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            return false;
+        }
+        previous = curr;
+    }
+
+    true
+}
+
+/// Checks if an iterator yields ordered values.
+pub fn is_sorted<T: Ord, I: Iterator<Item = T>>(mut iterator: I) -> bool {
```
Comment (zesterer) [inline bee-common/bee-common/src/ord.rs:26]: `iter.is_sorted()`?

Comment (thibault-martinez) [reply bee-common/bee-common/src/ord.rs:26]: It's nightly only

File: bee-common/bee-common/src/ord.rs (hunk 2 above)
Comment (zesterer) [reply bee-common/bee-common/src/ord.rs:16]: Eurgh

Comment (thibault-martinez) [reply bee-common/bee-common/src/ord.rs:16]: We'll remove when it's not anymore but in the meantime we kinda need this...

**Review by Alex6323 — APPROVED** [review]

File: bee-common/bee-common/src/ord.rs (hunk 1 above)
Comment (zesterer) [reply bee-common/bee-common/src/ord.rs:9]: From the name, I took it to mean "checks that every item in the iterator is unique from one-another"

Comment (thibault-martinez) [reply bee-common/bee-common/src/ord.rs:9]: That's what it does, or am I missing something ?

Comment (thibault-martinez) [reply bee-common/bee-common/src/ord.rs:9]: But since you also check if it's ordered, you don't need a hashmap, if that was your question

Comment (zesterer) [reply bee-common/bee-common/src/ord.rs:9]: > That's what it does, or am I missing something ?

Checks all elements against all others, I mean

Comment (thibault-martinez) [reply bee-common/bee-common/src/ord.rs:9]: Having an equal non-consecutive item would break the order so you don't need to do that

**Review by zesterer — APPROVED** [review]

**Pull request merged**